src/ExpressionParser.php

```diff
@@ -117,6 +118,25 @@ private function triggerPrecedenceDeprecations(AbstractExpression $expr, Token $
             ) {
                 trigger_deprecation('twig/twig', '3.15', \sprintf('As "+" / "-" will have a higher precedence than "~" in Twig 4.0, please add parentheses to keep the current behavior in "%s" at line %d.', $this->parser->getStream()->getSourceContext()->getName(), $token->getLine()));
             }
+
+            return;
+        }
+
+        // Precedence of the ?? operator will be lower than everything else in Twig 4.0
+        if ($expr instanceof NullCoalesceExpression) {
+            /** @var AbstractExpression $expr2 */
+            $expr2 = $expr->getNode('expr2');
+            /** @var AbstractExpression $expr3 */
+            $expr3 = $expr->getNode('expr3');
+            if (
+                (($expr2 instanceof AddBinary || $expr2 instanceof SubBinary || $expr2 instanceof ConditionalExpression) && !$expr2->hasExplicitParentheses())
+                ||
+                (($expr3 instanceof AddBinary || $expr3 instanceof SubBinary || $expr3 instanceof ConditionalExpression) && !$expr3->hasExplicitParentheses())
```

Comment:
Don't we have a lot more binary nodes that can be impacted (basically all of them as we go from the highest precedence among binary nodes to the lowest one).